### PR TITLE
ASL-780: Introduce collections

### PR DESCRIPTION
**Pull request by HadrienRenaud**
This PR introduces collections, a new type of values in ASL that functions
similarly to a record, but without considering them as a single element.

This makes no difference on the sequential semantics, but would allow herd+asl
to get rid of the 'experimental mode'.

The use of collections is strictly limited to global variables, and cannot be
accessed as a single element. This is implemented with 4 checks:
 1. A local storage element type cannot have the structure of a collection
 2. A function argument cannot be a collection
 3. A return type cannot be a collection
 4. ARBITRARY cannot use a collection type.

I think that this is enough to prevent un-implemented use of collections.

TODO:
- [x] Introduce `T_Collection` node in the AST
- [x] Give syntax for writing collections
- [x] Prevent the use of collection as local 
- [x] Add tests for use of collection outside of its boundaries
- [x] Implements concurrent semantics for collections
- [x] Add concurrent tests for collections

### CODE REVIEW

File: asllib/PP.ml [hunk 1]
```diff
@@ -206,16 +206,20 @@ and pp_ty f t =
...
+      fprintf f "@[array [[%a]] of %a@]" pp_array_length length pp_ty elt_type
+  | T_Collection record_ty -> pp_record_like f "collection" record_ty
+  | T_Record record_ty -> pp_record_like f "record" record_ty
+  | T_Exception record_ty -> pp_record_like f "exception" record_ty
   | T_Named x -> pp_print_string f x
 
+and pp_record_like f label record_ty =
+  fprintf f "@[<hv 2>%s {@ %a@;<1 -2>}@]" label pp_fields record_ty
+
+and pp_array_length f = function
```
Comment (Roman-Manevich) [inline asllib/PP.ml:219]: `pp_array_length` is a bit misleading, since what we want is the index kind. How about `pp_array_index` or `pp_index_kind`?

File: asllib/AST.mli [hunk 2]
```diff
@@ -252,6 +252,7 @@ and type_desc =
   | T_Array of array_index * ty
   | T_Record of field list
   | T_Exception of field list
+  | T_Collection of field list
```
Comment (Roman-Manevich) [inline asllib/AST.mli:255]: I wonder if abstracting over records/exceptions/collections would make things simpler. Right now there's a lot of duplicated code for record/exception/collection. This would also make the concept of "structured types" directly expressible:
```
| T_Structured of structured_kind * field list

and structured_kind = Record | Exception | Collection
```

**Review by Roman-Manevich — COMMENTED** [review]
See the comment suggesting abstracting record/exception/collection to a new AST node type.

**Review by Roman-Manevich — COMMENTED** [review]
Looks good so far. A few minor comments.

File: asllib/Interpreter.ml [hunk 3]
```diff
@@ -641,7 +641,22 @@ module Make (B : Backend.S) (C : Config) = struct
...
+    (* End *)
+    (* Begin EvalEGetCollectionFields *)
+    | E_GetCollectionFields (x, field_names) ->
+        let v_record = IEnv.find_global x env in
+        let scope = B.Scope.global ~init:false in
+        let* v_list =
+          List.map
+            (fun field_name ->
+              let* v = B.get_field field_name v_record in
+              let* () = B.on_read_identifier ((x ^ ".") ^ field_name) scope v in
```
Comment (Roman-Manevich) [inline asllib/Interpreter.ml:653]: Suggestion: move `(x ^ ".") ^ field_name` to its own function, e.g., `collection_field_effect_var`.

File: asllib/env.mli [hunk 4]
```diff
@@ -87,6 +87,9 @@ module type S = sig
   val find : identifier -> env -> v env_result
   (** Fetches an identifier from the environment. *)
 
+  val find_global : identifier -> env -> v
+  (** Fetches a global identifier. *)
```
Comment (Roman-Manevich) [inline asllib/env.mli:91]: `Fetches a global storage element`

File: asllib/doc/AssignableExpressions.tex [hunk 5]
```diff
@@ -1355,6 +1357,46 @@ \subsection{Typing}
...
+  \item $\vle$ denotes the access to the field named \texttt{field} in $\vleone$;
+  \item annotating the right-hand-side expression corresponding to $\vleone$ in $\tenv$ yields \\ $(\vtleone, \Ignore, \Ignore)$\ProseOrTypeError;
+  \item annotating the left-hand-side expression  $\vleone$ with type $\vtleone$ in $\tenv$ yields \\ $(\vletwo, \vses)$\ProseOrTypeError;
+  \item obtaining the \underlyingtype\ of $\vtleone$ in $\tenv$ yields a
+    collection type with fields $\fields$\ProseOrTypeError;
+  \item $\vletwo$ denotes a left-hand-side variable expression for $\vbase$, that is, $\LEVar(\vbase)$;
+  \item checking that there exists a type associated with the field $\field$ in $\fields$ $\True$ \ProseTerminateAs{\BadField};
+  \item the type associated with the field $\field$ in $\fields$ is $\vt$;
+  \item determining whether $\vte$ \typesatisfies\ $\vt$ yields $\True$\ProseOrTypeError;
+  \item $\newle$ is the access to the field $\field$ in $\vletwo$, that is, $\LESetCollectionFields(\vbase, [\field])$.
```
Comment (hrutvik) [inline asllib/doc/AssignableExpressions.tex:1381]:
```suggestion
  \item $\newle$ is the access to the field $\field$ in $\vletwo$, that is, $\LESetCollectionFields(\vletwo, [\field])$.
```

File: asllib/doc/AssignableExpressions.tex [hunk 6]
```diff
@@ -1401,6 +1443,24 @@ \subsection{Typing}
...
+        $\vbase$, that is, $\LEVar(\vbase)$;
+      \item applying $\foldbitvectorfields$ to $\vlefields$ and $\vbasefields$
+        in $\tenv$ yields $(\vlength, \vslices)$ \ProseOrTypeError;
+      \item \Proseeqdef{$\vtlhs$}{the bitvector type of length $\vlength$ and
+        no bitfields, that is, $\TBits(\ELInt{\vlength}, \emptylist)$};
+      \item checking that $\vte$ \typesatisfies\ $\vtlhs$ in $\tenv$ yields
+        $\True$\ProseOrTypeError;
+      \item define $\newle$ as the assignable expression of the list of fields
+        $\vlefields$ to the collection base expression $\vbase$, that is, \\
+        $\LESetCollectionFields(\lebase, \vlefields)$;
```
Comment (hrutvik) [inline asllib/doc/AssignableExpressions.tex:1460]:
```suggestion
        $\LESetCollectionFields(\vbase, \vlefields)$;
```

File: asllib/ASTUtils.ml [hunk 7]
```diff
@@ -773,6 +784,7 @@ let rename_locals map_name ast =
     | LE_SetEnumArray (le, i) -> LE_SetEnumArray (map_le le, map_e i)
     | LE_SetField (le1, f) -> LE_SetField (map_le le1, f)
     | LE_SetFields (le1, fl, annot) -> LE_SetFields (map_le le1, fl, annot)
+    | LE_SetCollectionFields _ as le -> le (* No collection is local *)
```
Comment (Roman-Manevich) [inline asllib/ASTUtils.ml:787]: I don't understand the comment `(* No collection is local *)`

File: asllib/doc/AssignableExpressions.tex [hunk 8]
```diff
@@ -1633,6 +1710,48 @@ \subsection{Semantics}
...
+    of the collection global storage element $\vbase$ with the corresponding
+    slices given in $\vslices$ from the bitvector value $\vv$ (the rule
+    \TypingRuleRef{LESetFields} ensures that the length of $\fields$ and
+    $\vslices$ is the same);
+  \item view $\env$ as an environment where $\tenv$ is the static environment
+    and $\denv$ is the dynamic environment;
+  \item $\vbase$ is bound in the global dynamic environment ($G^\denv.\storage$);
+  \item $\vvone$ is the value of $\vbase$ in the global component of $\env$ ;
+  \item view $\vm$ as $(\vv, \vg)$;
+  \item view $\vrmrecord$ as $(\vvone, \{\})$;
```
Comment (hrutvik) [inline asllib/doc/AssignableExpressions.tex:1737]: Is `\vm` defined? It looks like a free variable
And then should `\vmrecord` use `\Prosedefine` rather than "view"?

File: asllib/tests/ASLTypingReference.t/TypingRule.CheckIsNotCollection.asl [hunk 9]
```diff
@@ -0,0 +1,13 @@
+type MyCollection of collection {
+  field1: bits(2),
+  field2: bits(3),
+};
+
+func main () => integer
+begin
+  var test: MyCollection;
```
Comment (Roman-Manevich) [inline asllib/tests/ASLTypingReference.t/TypingRule.CheckIsNotCollection.asl:8]: // Illegal: local storage elements cannot have collection types.

File: asllib/doc/Types.tex [hunk 10]
```diff
@@ -1350,6 +1351,30 @@ \subsection{Typing Record Types\label{sec:TypingRecordTypes}}
 \end{mathpar}
 \CodeSubsection{\TStructuredDeclBegin}{\TStructuredDeclEnd}{../Typing.ml}
 
+\section{Collection Types\label{sec:CollectionTypes}}%
+\hypertarget{collectiontypeterm}{}
+
+A collection is a \structuredtype{} consisting of a list of field identifiers
+which denote individual storage elements.
+
```
Comment (Roman-Manevich) [inline asllib/doc/Types.tex:1359]: Consider adding
```
\RequirementRule{CollectionsGlobal}
Collections can only be used for global storage elements. See \listingref{checkisnotcollection} for an ill-typed specification.
```

**Review by Roman-Manevich — APPROVED** [review]

**Review by hrutvik — APPROVED** [review]
